lib/rubocop/cop/lint/mixed_regexp_capture_types.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Do not mix named captures and numbered captures in a Regexp literal
+      # because numbered capture is ignored if they're mixed.
+      # Replace numbered captures with non-capturing groupings or
+      # named captures.
+      #
+      #   # bad
+      #   /(?<foo>FOO)(BAR)/
+      #
+      #   # good
+      #   /(?<foo>FOO)(?<bar>BAR)/
+      #
+      #   # good
+      #   /(?<foo>FOO)(?:BAR)/
+      #
+      class MixedRegexpCaptureTypes < Cop
+        MSG = 'Do not mix named captures and numbered captures ' \
+              'in a Regexp literal.'
+
+        def on_regexp(node)
+          tree = Regexp::Parser.parse(node.content)
+          return unless named_capture?(tree)
+          return unless numbered_capture?(tree)
+
+          add_offense(node)
+        end
+
+        private
+
+        def named_capture?(tree)
+          each_regexp_expr(tree).any? do |e|
+            e.instance_of?(Regexp::Expression::Group::Capture)
+          end
+        end
+
+        def numbered_capture?(tree)
+          each_regexp_expr(tree).any? do |e|
+            e.instance_of?(Regexp::Expression::Group::Named)
+          end
+        end
+
+        def each_regexp_expr(tree)
+          tree.to_enum(:each_expression)
```

Comment:
I'll refactor it after the new version of regexp_parser gem is released. ref: https://github.com/ammar/regexp_parser/pull/62